internal/sbi/consumer/nrf_service.go

```diff
@@ -0,0 +1,302 @@
+		return ""
+	}
+	for _, profile := range result.NfInstances {
+		if uri := util.SearchNFServiceUri(profile, models.ServiceName_NBSF_MANAGEMENT,
+			models.NfServiceStatus_REGISTERED); uri != "" {
+			return uri
+		}
+	}
+	return ""
+}
+
+func (s *nnrfService) SendNFInstancesAMF(nrfUri string, guami models.Guami, serviceName models.ServiceName) string {
+	targetNfType := models.NfType_AMF
+	requestNfType := models.NfType_PCF
+
+	localVarOptionals := Nnrf_NFDiscovery.SearchNFInstancesParamOpts{
+		Guami: optional.NewInterface(util.MarshToJsonString(guami)),
+	}
+	// switch types {
+	// case NFDiscoveryToUDRParamSupi:
+	// 	localVarOptionals.Supi = optional.NewString(id)
+	// case NFDiscoveryToUDRParamExtGroupId:
+	// 	localVarOptionals.ExternalGroupIdentity = optional.NewString(id)
+	// case NFDiscoveryToUDRParamGpsi:
+	// 	localVarOptionals.Gpsi = optional.NewString(id)
+	// }
+
+	result, err := s.SendSearchNFInstances(nrfUri, targetNfType, requestNfType, localVarOptionals)
+	if err != nil {
+		logger.ConsumerLog.Error(err.Error())
+		return ""
+	}
+	for _, profile := range result.NfInstances {
+		return util.SearchNFServiceUri(profile, serviceName, models.NfServiceStatus_REGISTERED)
+	}
+	return ""
+}
+
+// management
+func (s *nnrfService) BuildNFInstance(context *pcf_context.PCFContext) (profile models.NfProfile, err error) {
+	profile.NfInstanceId = context.NfId
+	profile.NfType = models.NfType_PCF
+	profile.NfStatus = models.NfStatus_REGISTERED
+	profile.Ipv4Addresses = append(profile.Ipv4Addresses, context.RegisterIPv4)
+	service := []models.NfService{}
+	for _, nfService := range context.NfService {
+		service = append(service, nfService)
+	}
+	profile.NfServices = &service
+	profile.PcfInfo = &models.PcfInfo{
+		DnnList: []string{
+			"free5gc",
+			"internet",
+		},
+		// SupiRanges: &[]models.SupiRange{
+		// 	{
+		// 		//from TS 29.510 6.1.6.2.9 example2
+		//		//no need to set supirange in this moment 2019/10/4
+		// 		Start:   "123456789040000",
+		// 		End:     "123456789059999",
+		// 		Pattern: "^imsi-12345678904[0-9]{4}$",
+		// 	},
+		// },
+	}
+	if context.Locality != "" {
+		profile.Locality = context.Locality
+	}
+	return
+}
+
+func (s *nnrfService) SendRegisterNFInstance(ctx context.Context) (
+	resouceNrfUri string, retrieveNfInstanceID string, err error,
+) {
+	// Set client and set url
+	pcfContext := s.consumer.Context()
+
+	client := s.getNFManagementClient(pcfContext.NrfUri)
+	nfProfile, err := s.BuildNFInstance(pcfContext)
+	if err != nil {
+		return "", "",
+			errors.Wrap(err, "RegisterNFInstance buildNfProfile()")
+	}
+
+	var nf models.NfProfile
+	var res *http.Response
+	for {
+		nf, res, err = client.NFInstanceIDDocumentApi.RegisterNFInstance(context.TODO(), pcfContext.NfId, nfProfile)
+		if err != nil || res == nil {
+			logger.ConsumerLog.Errorf("CHF register to NRF Error[%v]", err)
+			time.Sleep(2 * time.Second)
+			continue
+		}
+		defer func() {
+			if resCloseErr := res.Body.Close(); resCloseErr != nil {
```

Comment:
Don't close the body twice. The API client will close it after parsing. [See here](https://github.com/free5gc/openapi/blob/main/Nnrf_NFManagement/api_nf_instance_id_document.go#L308)